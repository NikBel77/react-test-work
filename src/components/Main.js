import React from 'react'

import DataTable from './DataTable'
import PersonInfo from './PersonInfo'
import ControlPanel from './ControlPanel'
import Pagination from './Pagination'

import { paginationFilter, filterByStr, sortByColumnName } from '../utils'
import { ROWS_ON_TABLE } from '../constants'

class Main extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            data: [],
            selectedPerson: false,

            columnSort: {
                columnName: 'id',
                isReverseDirection: false,
            },

            searchStr: '',
            currentPage: 1,
        };


        this.addNewPerson = this.addNewPerson.bind(this);
        this.changeActiveColumn = this.changeActiveColumn.bind(this);
    }

    changeState(prop, value) {
        if(!value && prop !== 'searchStr') return;
        if(prop === 'data') {
            this.setState({
                data: value,
                searchStr: '',
                currentPage: 1,
            })
        } else {
            this.setState({
                [prop]: value,
            })
        }
    }

    changeActiveColumn(columnName) {
        if (this.state.columnSort.columnName === columnName) {
            const direction = !this.state.columnSort.isReverseDirection;
            this.setState({
                columnSort: {
                    columnName,
                    isReverseDirection: direction,
                }
            })
        } else {
            this.setState({
                columnSort: {
                    columnName,
                    isReverseDirection: false,
                }
            })
        }
    }

    addNewPerson(person) {
        const data = this.state.data;
        data.push(person);
        this.setState({
            data,
        })
    }

    mainFilter(data) {
        if(!data) throw new Error('empty Data')
        if(!data.length) return data

        let sortedData = sortByColumnName
            (
                filterByStr(data, this.state.searchStr),
                this.state.columnSort.columnName,
                this.state.columnSort.isReverseDirection
            );

        return sortedData
    }

    render() {
        const sortedData = this.mainFilter(this.state.data);

        const numberOfPages = Math.floor(sortedData.length / ROWS_ON_TABLE) 
            + ((sortedData.length % ROWS_ON_TABLE) > 0 ? 1 : 0);

        const resCurrentPage = this.state.currentPage <= numberOfPages
            ? this.state.currentPage
            : numberOfPages;
        const resData = paginationFilter(sortedData, resCurrentPage, ROWS_ON_TABLE);


        return (
            <main>

                <ControlPanel
                    onNewPerson={this.addNewPerson}
                    onDataLoaded={(value) => this.changeState('data', value)}
                    onSearch={(value) => this.changeState('searchStr', value)}
                />

                <hr />

                <Pagination 
                    numberOfPages={numberOfPages}
                    currentPage={resCurrentPage}
                    onPageChange={(value) => this.changeState('currentPage', value)}
                />

                <DataTable
                    dataList={resData}
                    onSelectPerson={(value) => this.changeState('selectedPerson', value)}
                    activePerson={this.state.selectedPerson}
                    activeColumn={this.state.columnSort}
                    onChangeActiveColumn={this.changeActiveColumn}
                />

                <hr />

                <PersonInfo personInfo={this.state.selectedPerson} />

            </main>
        )
    }
}

export default Main;