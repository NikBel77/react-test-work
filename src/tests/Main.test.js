import React from 'react'
import ReactDOM from 'react-dom'
import Main from '../components/Main'
import renderer from 'react-test-renderer'

test('render without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Main />, div);
    ReactDOM.unmountComponentAtNode(div);
});
test('snapshot', () => {
    const tree = renderer.create(<Main />).toJSON();
    expect(tree).toMatchSnapshot();
});