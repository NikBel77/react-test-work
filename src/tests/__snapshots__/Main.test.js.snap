// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<main>
  <div
    className="container"
  >
    <div
      className="control-panel"
    >
      <div
        className="control-panel__left"
      >
        <div
          className="demo-loader"
        >
          <button
            className="btn btn-success"
            onClick={[Function]}
          >
            short demo list
          </button>
          <button
            className="btn btn-danger"
            onClick={[Function]}
          >
            long demo list
          </button>
        </div>
        <div
          className="search-panel"
        >
          <input
            type="text"
          />
          <button
            className="btn btn-primary"
            onClick={[Function]}
          >
            Search
          </button>
        </div>
      </div>
      <div
        className="control-panel__right"
      >
        <div>
          <button
            className="btn btn-light"
            onClick={[Function]}
          >
            <img
              alt="plus"
              src="plus.svg"
            />
          </button>
          <div
            style={
              Object {
                "display": "none",
              }
            }
          >
            <form
              className="add-form"
            >
              <div
                className="form-row"
              >
                <div
                  className="form-group col-md-1"
                >
                  <label
                    htmlFor="inputId"
                  >
                    ID
                  </label>
                  <input
                    className="form-control"
                    data-name="id"
                    id="inputId"
                    placeholder="id"
                    type="text"
                  />
                </div>
                <div
                  className="form-group col-md-3"
                >
                  <label
                    htmlFor="inputName"
                  >
                    First name
                  </label>
                  <input
                    className="form-control"
                    data-name="firstName"
                    id="inputName"
                    placeholder="enter name"
                    type="text"
                  />
                </div>
                <div
                  className="form-group col-md-3"
                >
                  <label
                    htmlFor="inputLastName"
                  >
                    Last name
                  </label>
                  <input
                    className="form-control"
                    data-name="lastName"
                    id="inputLastName"
                    placeholder="enter last name"
                    type="text"
                  />
                </div>
                <div
                  className="form-group col-md-3"
                >
                  <label
                    htmlFor="inputEmail"
                  >
                    Email
                  </label>
                  <input
                    className="form-control"
                    data-name="email"
                    id="inputEmail"
                    placeholder="email"
                    type="email"
                  />
                </div>
                <div
                  className="form-group col-md-2"
                >
                  <label
                    htmlFor="inputPhone"
                  >
                    Phone
                  </label>
                  <input
                    className="form-control"
                    data-name="phone"
                    id="inputPhone"
                    placeholder="phone"
                    type="text"
                  />
                </div>
              </div>
              <button
                className="btn btn-primary"
                onClick={[Function]}
              >
                Add
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
  <hr />
  <div
    className="container"
  >
    <table
      className="table table-hover table-sm table-dark"
    >
      <thead>
        <tr>
          <th
            onClick={[Function]}
            scope="col"
          >
            ID
            <img
              alt="caret"
              className="table-caret"
              src="caret-d.svg"
            />
          </th>
          <th
            onClick={[Function]}
            scope="col"
          >
            First name
            <img
              alt="caret"
              className="table-caret"
              src="caret-r.svg"
            />
          </th>
          <th
            onClick={[Function]}
            scope="col"
          >
            Last name
            <img
              alt="caret"
              className="table-caret"
              src="caret-r.svg"
            />
          </th>
          <th
            onClick={[Function]}
            scope="col"
          >
            Email
            <img
              alt="caret"
              className="table-caret"
              src="caret-r.svg"
            />
          </th>
          <th
            onClick={[Function]}
            scope="col"
          >
            Phone
            <img
              alt="caret"
              className="table-caret"
              src="caret-r.svg"
            />
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            colSpan="5"
          >
            no data
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <hr />
  <div
    className="container"
  >
    <p>
      no user selected
    </p>
  </div>
</main>
`;
